nntrainer/src/databuffer_v2.cpp

```diff
@@ -0,0 +1,325 @@
+    buffer_cond_hungry.notify_all();
+
+    /** wait for thread to stop */
+    buffer_cond_filled.wait(thread_stop, [this] {
+      return thread_state != ThreadStates::THREAD_REQUEST_TO_STOP;
+    });
+  }
+
+  /** wait for thread to join */
+  joinDataCollectors();
+
+  /** throw error in case of error */
+  if (thread_state == ThreadStates::THREAD_ERROR)
+    throw std::logic_error("Dataset background thread resulted in error");
+
+  thread_state = ThreadStates::THREAD_READY;
+  buffer.clear();
+  batched_buffer.clear();
+  avail_buffer_idx = 0;
+}
+
+void DataBuffer_v2::start() {
+  {
+    /**
+     * Lock is needed here as the data buffer might already be running
+     * or the stop called hasnt been finished
+     */
+    std::lock_guard<std::mutex> thread_start(thread_m);
+
+    if (thread_state == ThreadStates::THREAD_NULL)
+      throw std::runtime_error("starting an un-initilaized dataset");
+    else if (thread_state == ThreadStates::THREAD_ERROR)
+      throw std::runtime_error("starting a data buffer which is in error");
+    else if (thread_state != ThreadStates::THREAD_READY)
+      throw std::runtime_error("starting an already running data buffer");
+
+    runDataCollectors();
+    thread_state = ThreadStates::THREAD_RUNNING;
+  }
+
+  /** create buffer of size buffer_len */
+  for (size_t batched_buflen = 0; batched_buflen < batched_buffer_len;
+       batched_buflen++) {
+    pushBatchedData();
+
+    /** As soon as there are more data containers available, start the data
+     * collectors */
+    // TODO: move this notify inside the pushBatchedData() itself.
+    buffer_cond_hungry.notify_all();
+  }
+}
+
+int DataBuffer_v2::getData(std::vector<sharedTensor> &inputs,
+                           std::vector<sharedTensor> &labels) {
+  if (inputs.size() != input_size.size() || labels.size() != label_size.size())
+    throw std::invalid_argument(
+      "Number of inputs and labels mismatch with the set configuration");
+
+  /** Ensure that the dataset is actually running */
+  {
+    std::lock_guard<std::mutex> thread_start(thread_m);
+    if (thread_state != ThreadStates::THREAD_RUNNING &&
+        thread_state != ThreadStates::THREAD_EPOCH_FINISHED)
+      throw std::runtime_error("Calling get data on a thread not running");
+  }
+
+  /** Keep on looping till a batch of data is ready */
+  while (true) {
+    {
+      std::lock_guard<std::mutex> thread_start(thread_m);
+      if (thread_state == ThreadStates::THREAD_EPOCH_FINISHED)
+        return false;
+      else if (thread_state != ThreadStates::THREAD_RUNNING)
+        throw std::logic_error("Background thread is not running anymore");
+    }
+
+    bool data_available;
+    /** check if data is available */
+    {
+      std::lock_guard<std::mutex> batched_buffer_data_avail_check_lock(
+        batched_buffer_m);
+      auto batched_buffer_elem = batched_buffer.front();
+      data_available = std::get<2>(batched_buffer_elem) == batch_size;
```

Comment:
It seems like we don't need to use batched_buffer if we check `data_available` by available_buffer_idx > batch_size making code and locks simpler?